libraries/Microsoft.Bot.Connector/OAuthApiClient.cs

```diff
@@ -0,0 +1,350 @@
+                tracingParameters.Add("userId", userId);
+                tracingParameters.Add("connectionName", connectionName);
+                tracingParameters.Add("magicCode", magicCode);
+                tracingParameters.Add("cancellationToken", cancellationToken);
+                ServiceClientTracing.Enter(invocationId, this, "GetUserTokenAsync", tracingParameters);
+            }
+            // Construct URL
+            var tokenUrl = new Uri(new Uri(uri + (uri.EndsWith("/") ? "" : "/")), "api/usertoken/GetToken?userId={userId}&connectionName={connectionName}{magicCodeParam}").ToString();
+            tokenUrl = tokenUrl.Replace("{connectionName}", Uri.EscapeDataString(connectionName));
+            tokenUrl = tokenUrl.Replace("{userId}", Uri.EscapeDataString(userId));
+            if (!string.IsNullOrEmpty(magicCode))
+            {
+                tokenUrl = tokenUrl.Replace("{magicCodeParam}", $"&code={Uri.EscapeDataString(magicCode)}");
+            }
+            else
+            {
+                tokenUrl = tokenUrl.Replace("{magicCodeParam}", String.Empty);
+            }
+
+            // Create HTTP transport objects
+            var httpRequest = new HttpRequestMessage();
+            HttpResponseMessage httpResponse = null;
+            httpRequest.Method = new HttpMethod("GET");
+            httpRequest.RequestUri = new Uri(tokenUrl);
+
+            // add botframework api service url to the list of trusted service url's for these app credentials.
+            MicrosoftAppCredentials.TrustServiceUrl(tokenUrl);
+
+            // Set Credentials
+            if (client.Credentials != null)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                await client.Credentials.ProcessHttpRequestAsync(httpRequest, cancellationToken).ConfigureAwait(false);
+            }
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (shouldTrace)
+            {
+                ServiceClientTracing.SendRequest(invocationId, httpRequest);
+            }
+            httpResponse = await client.HttpClient.SendAsync(httpRequest, cancellationToken).ConfigureAwait(false);
+            if (shouldTrace)
+            {
+                ServiceClientTracing.ReceiveResponse(invocationId, httpResponse);
+            }
+            HttpStatusCode statusCode = httpResponse.StatusCode;
+            cancellationToken.ThrowIfCancellationRequested();
+            if (statusCode == HttpStatusCode.OK)
+            {
+                string responseContent = await httpResponse.Content.ReadAsStringAsync().ConfigureAwait(false);
+                try
+                {
+                    var tokenResponse = Rest.Serialization.SafeJsonConvert.DeserializeObject<TokenResponse>(responseContent);
+                    return tokenResponse;
+                }
+                catch (JsonException)
+                {
```

Comment:
This exception handling seems... incorrect. If it's right, please add comments in here.